drivers/led/lp50xx.c

```diff
@@ -330,6 +331,34 @@ static int lp50xx_init(const struct device *dev)
 	return 0;
 }
 
+#ifdef CONFIG_PM_DEVICE
+static int lp50xx_pm_action(const struct device *dev,
+			    enum pm_device_action action)
+{
+	int err;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		return lp50xx_enable(dev, false);
+	case PM_DEVICE_ACTION_RESUME:
+		return lp50xx_enable(dev, true);
+	case PM_DEVICE_ACTION_TURN_OFF:
+		return lp50xx_hw_enable(dev, false);
+	case PM_DEVICE_ACTION_TURN_ON:
+		err = lp50xx_hw_enable(dev, true);
```

Comment:
`gpio_enable` is not required, so returning 0 seems the right thing to do when it is not present in the devicetree.

Reply:
I am not familiar with the PM API but I think there is a misunderstanding here. `TURN_ON` and `TURN_OFF` actions are only triggered by a power domain when it is resumed or suspended. So the device/driver can prepare for the power to be removed/added. That's what not we are doing here.

So eventually we should move `gpio_enable` under the `SUSPEND` / `RESUME` action handling and return 0 for  `TURN_ON` / `TURN_OFF` actions (because there is indeed nothing to do). the  What do you think ?

Reply:
Maybe I can use the same behavior as `lp5569`, i.e. use the same action for `TURN_ON`/`RESUME` and `TURN_OFF`/`SUSPEND`, but configuration will be lost each time if I don't keep a copy of the registers to restore inside `lp50xx_data`, am I right ?

Reply:
No the implementation of the `lp5569` driver doesn't look good. See the "Device actions x states" diagram on https://docs.zephyrproject.org/latest/services/pm/device.html. The `SUSPEND` action is always executed before the `TURN_OFF` action. So the `lp5569` driver runs the same code twice to reach the `OFF` state... Same goes to reach the `ON` state. So I believe it is correct to ignore the `TURN_ON` and `TURN_OFF` PM action in a LED driver.

With your original implementation, `gpio_enable` is basically useless because it needs a power domain to go off (which is unlikely to happen).

But if you use `gpio_enable` when suspending the device, then you are indeed in trouble to restore the registers when resuming. In addition, suspending a device is actually quite different than powering it off...

There is another thing to consider: some apps may not want the LEDs to go off when the device is suspended. An application could even use the LEDs to notify users about the power status. But unfortunately there is no API to disable PM support for a given device/driver...

That's why in https://github.com/zephyrproject-rtos/zephyr/pull/56673, @danieldegrasse choose to implement blanking support through a private API. See the discussion https://github.com/zephyrproject-rtos/zephyr/pull/56673/commits/99bb95315b84f5a3da8bb0a5017d169667dd66ed#r1250848989. Even if I don't like much the idea, I had nothing better to suggest. Maybe we could do the same here with `gpio_enable` ?

Reply:
Hi @ceolin and @gmarull. Please can you offer some guidance here ?